### ISSUE TITLE
Regression with ansible_pkg_mgr discovery on Amazon Linux 2

### ISSUE THREAD

**Issue by acdha** [ansible/ansible#83428]
### Summary

With Ansible 9, the package manager on an Amazon Linux 2 system is correctly detected as `yum`. With Ansible 10, it is detected as “unknown”.

This is compounded due to a separate bug in the `dnf` module which causes it to override an explicit `use_backend: yum` directive with the original `unknown` value:

https://github.com/ansible/ansible/blob/502241588787a6ea719813f378378c01557e2f56/lib/ansible/plugins/action/dnf.py#L29-L38

The end result is that with configuration like this:

```yaml
      ansible.builtin.dnf:
        use_backend: yum
        name: "*"
        state: latest
        update_cache: true
        update_only: true
```

… systems which work with Ansible 9 will fail with an error like this:

```
fatal: [10.52.152.19]: FAILED! => {"ansible_facts": {"pkg_mgr": "unknown"}, "changed": false, "msg": ["Could not detect which major revision of dnf is in use, which is required to determine module backend.", "You should manually specify use_backend to tell the module whether to use the dnf4 or dnf5 backend})"]}
```

### Issue Type

Bug Report

### Component Name

setup

### Ansible Version

```console
ansible [core 2.17.0]
  config file = /Users/cadams/Projects/…/ansible.cfg
  configured module search path = ['/Users/cadams/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /opt/homebrew/Cellar/ansible/10.0.1/libexec/lib/python3.12/site-packages/ansible
  ansible collection location = /Users/cadams/.ansible/collections:/usr/share/ansible/collections
  executable location = /opt/homebrew/bin/ansible
  python version = 3.12.3 (main, Apr  9 2024, 08:09:14) [Clang 15.0.0 (clang-1500.3.9.4)] (/opt/homebrew/Cellar/ansible/10.0.1/libexec/bin/python)
  jinja version = 3.1.4
  libyaml = True
```

### Configuration

```console
# if using a version older than ansible-core 2.12 you should omit the '-t all'
$ ansible-config dump --only-changed -t all
CONFIG_FILE() = /Users/cadams/Projects/…/ansible.cfg
DEFAULT_VAULT_PASSWORD_FILE(/Users/cadams/Projects/…/ansible.cfg) = /Users/cadams/Projects/…/get_vault_password_from_keyring
EDITOR(env: EDITOR) = vim
INTERPRETER_PYTHON(/Users/cadams/Projects/…/ansible.cfg) = /usr/bin/python3
INVENTORY_ENABLED(/Users/cadams/Projects/…/ansible.cfg) = ['aws_ec2']
PERSISTENT_COMMAND_TIMEOUT(/Users/cadams/Projects/…/ansible.cfg) = 60
PERSISTENT_CONNECT_RETRY_TIMEOUT(/Users/cadams/Projects/…/ansible.cfg) = 45
PERSISTENT_CONNECT_TIMEOUT(/Users/cadams/Projects/…/ansible.cfg) = 90
USE_PERSISTENT_CONNECTIONS(/Users/cadams/Projects/…/ansible.cfg) = True

CONNECTION:
==========

paramiko_ssh:
____________
ssh_args(/Users/cadams/Projects/…/ansible.cfg) = -o ControlMaster=auto -o ControlPersist=600s
use_persistent_connections(/Users/cadams/Projects/…/ansible.cfg) = True

ssh:
___
pipelining(/Users/cadams/Projects/…/ansible.cfg) = True
reconnection_retries(/Users/cadams/Projects/…/ansible.cfg) = 3
ssh_args(/Users/cadams/Projects/…/ansible.cfg) = -o ControlMaster=auto -o ControlPersist=600s
```

### OS / Environment

Amazon Linux 2

### Steps to Reproduce

```yaml (paste below)
      ansible.builtin.dnf:
        use_backend: yum
        name: "*"
        state: latest

```

### Expected Results

The YUM package manager is used as configured or the [`ansible.builtin.dnf` documentation](https://docs.ansible.com/ansible/latest/collections/ansible/builtin/dnf_module.html#parameter-use_backend) would not state it supports `use_backend: yum` 

### Actual Results

```console
fatal: [10.52.152.19]: FAILED! => {"ansible_facts": {"pkg_mgr": "unknown"}, "changed": false, "msg": ["Could not detect which major revision of dnf is in use, which is required to determine module backend.", "You should manually specify use_backend to tell the module whether to use the dnf4 or dnf5 backend})"]}
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

**Comment by ansibot** [comment]
Files identified in the description:

* [`lib/ansible/modules/setup.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/setup.py)

If these files are incorrect, please update the `component name` section of the description or use the [component](https://github.com/ansible/ansibotmini#commands) bot command.

**Comment by flowerysong** [comment]
This is a deliberate removal of functionality, not a regression. As noted in the [porting guide](https://docs.ansible.com/ansible/latest/porting_guides/porting_guide_10.html#id6) Python 2 and the original `yum` module are no longer supported. If you need to manage systems like Amazon Linux 2 which are not compatible with this version of ansible-core, you should downgrade to a version that does support these targets.

`use_backend: yum` is supported to (theoretically) allow people to avoid updating their playbooks, though I can't think of a situation where a playbook with `yum: use_backend=yum` would have worked in the past and continue working now, so it arguably should have just been removed. Since it no longer maps to a specific supported backend the logic falls back to other methods of determining the appropriate backend, and there is no backend that works with AL2 so it's expected that it fails.

**Comment by acdha** [comment]
> This is a deliberate removal of functionality, not a regression. As noted in the [porting guide](https://docs.ansible.com/ansible/latest/porting_guides/porting_guide_10.html#id6) Python 2 and the original `yum` module are no longer supported. If you need to manage systems like Amazon Linux 2 which are not compatible with this version of ansible-core, you should downgrade to a version that does support these targets.

This project is using Python 3 (see above) and the documentation for the current version of the **DNF** module being used (see above) says that the `yum` backend is a valid choice.

**Comment by flowerysong** [comment]
It's a valid choice, but it doesn't do the same thing it used to, which is select the `yum` module.. The `yum` module only worked with Python 2 (because that's the only version the Python bindings in AL2 are built for, regardless of what Python versions are installed on the system) so it has been removed.

**Comment by acdha** [comment]
In that case, there is the original bug (the package manager shouldn't be detected as “unknown”) and the documentation should be updated to remove the unsupported yum option, along with updating the porting guide to note that this removal affects more than just the yum module.

**Comment by flowerysong** [comment]
#83429 updates the documentation for `use_backend` to be much clearer about what each option means.

**Comment by sivel** [comment]
Resolved by https://github.com/ansible/ansible/pull/83429

**Comment by acdha** [comment]
> The package manager is detected as "unknown" because no supported package manager is found. This is not a bug.

I think updating the documentation to say that Ansible 10 does not support the EL7-era distributions even when running under Python 3 is the right fix but this is a bug by common definitions of the term: Ansible 9 returns the correct output, Ansible 10 does not, and the cause not due to any change to those servers – from a user's perspective, the identical input is now returning an incorrect output.  Anyone using Ansible with logic based on the discovered facts will now get incorrect results even for code which would otherwise work (e.g. I have used that fact to run certain configuration `command` blocks for things which were never supported by the yum module).

I think the update to the documentation is probably sufficient but it might be worth clarifying some of the vagueness: the porting guide and ansible-core's support matrix do say that Python 3.7+ is required but all of those servers I encountered this on have Python 3.8 installed as `/usr/bin/python3`. What they don't have are the Python bindings which the `yum` RPM provided for Python 2.7 on distributions of that era. Perhaps the porting guide should explicitly say that the `yum` functionality not available even when using an otherwise-supported Python version – as currently written, it sounds like a system where Ansible is using `/usr/bin/python3.8` would work, or potentially require switching to the `dnf` module since “redirected” is generally not a term you'd used to say something has been removed without a migration path. 

I think it would be useful to make the unsupported aspect easy to discover quickly since package management is such a common use for Ansible and many large organizations will have legacy systems which need to be patched, especially since some distributions will be supported for at least a year from now. I think it would be useful to let people know as early as possible that they'll need to keep Ansible 9 around until those are decommissioned.
